### DISCUSSION TITLE
Testing User Session with cookies() in Playwright + NextJS

### DISCUSSION THREAD

**Discussion by GarethRader** [vercel/next.js#62254] [Help]
### Summary

I have routes in my NextJS 14.0.4 app that are protected via user session/JWT that are retrieved from a `login` server action. This all works well, and I implemented everything according to the docs. However, writing any sort of user flow integration/e2e tests that deals with cookies does not seem to work.

The issue I am running into seems to be that NextJS App cannot use Playwright's `context().addCookies()` and Playwright is not able to use `next/headers` `cookies()` methods. This causes any test that tests/verifies cookies for server side rendering/logic to error.

In my app specifically, the protected routes get automatically redirected despite waiting for the server action post request response (and it successfully completing). In addition, I have certain UI components that are conditionally rendered based on whether or not a user session is valid/exists (logout button in header component etc...). These UI components do not render in the automated playwright tests whether I add the cookies in the playwright test, nor if I have the test fill form inputs -> click login button -> await post request response -> redirect to page with playwright's `page.goto()`. 

The app itself is a login form, main page, and a header component used in root `layout.tsx` that conditionally renders a logout button depending on the user session. Main page redirects to `/auth` login page if no session exists or is invalid. There is also a `login` server action linked to the login form that creates the session, and a `logout` server action linked to the logout button that just deletes the cookie.

All of my attempts to directly/indirectly call `next/headers` `cookies()` with a testing util function or otherwise usually gets a `invariant: cookies() expects to have requestasyncstorage none available ` error despite wrapping parent functions and such in async/await etc... Logging in both nextJS app and playwright show that neither can use the other's cookies as they are null values.

I have not found any Github Issues, stack overflow posts, or example documentation in NextJS or Playwright that cover this use case specifically. It would be very helpful to know if someone has found a solution, workaround that they can share or verify that this is a problem they have run into as well.

Please let me know if a provided reproducible code example would be more helpful. I cannot share the exact code due to company policy, however I will make an example I can share that showcases the issue I am running into. 

```
// layout.tsx

export default function RootLayout({
children}: {children: React.ReactNode}){
 return (
  <html>
   <body>
    <Header/>
    <main>{children}</main>
   </body>
  </html>
 )
}

```

```
// page.tsx

export default async function Home(){
 if (!(await getSession()){
   redirect("/auth") // have troubles with redirects in middleware.ts so put protection here
 }
 
 return (
  <div>
   <MainPage/>
  </div>
 )
}
```

```
// components/Header/Header.tsx

export default async function Header(){
 return (
  <div>
  <Image src={logo}/>
  <h1>My App Name </h1>
  {(await getSession()) && ( <Logout />)}
  </div>
 )
}
```

```
// components/Header/Logout,tsx
'use client'

const Logout = () => {
const {pending} = useFormStatus()
 return (
  <form action={}>
   <Button type="submit>
   Logout
   </Button>
  </form>
 )
}

export default Logout
```

```
// auth/page.tsx
'use client'

const Submit = ()=>{
 const {pending} = useFormStatus()
 return (
 <Button
  type="submit"
  aria-disabled={pending}
 >
    Login
 </Button>
 )
}
const initialState = { status: ""}


const Auth = ()=>{
  const [state, formAction] = useFormState(Login, initialState)
  return (
   <form action={formAction}>
      // MUI components
      <TextField
         name="username"
         label="username"
      />
      <TextField
         name="password"
         label="password"
         type="password
      />
      <Submit/>
   </form>
  )
}

export default Auth
```

```
// action/login.ts
export default async function Login(prevState: any, formData: FormData){
 const params = {
  username: formData.get("username"),
  password: formData.get("password")
 }

 try{
  const login = await fetch(process.env.ENDPOINT, {
       method: "POST",
       cache: "no-cache",
       headers: {
         "Content-Type": "application/json"
       }
       body: JSON.stringify(params)
     }
  const data = await login.json()

   await cookies().set({
     name: "app-session",
     value: data.token,
     httpOnly: true,
     secure: process.env.NODE_ENV === "production"
   })
 } catch (err) {
   console.error(err)
   return {status: "error", type "INVALID_AUTH"}
 }

 //success
 return redirect("/")
}
```

```
// playwright/test.spec.ts

test("UI renders correctly", async ({page})=>{
 await page.context().addCookies([{
     name: "app-session",
     value: params[0].token, //loaded from .env
     httpOnly: true,
     secure: false,
     path: "/",
     domain: "localhost:3000" // have also tried "http://localhost:3000"
     
   await page.goto("/")
   
   //automatically redirects to /auth route
   // If redirect in root level page.tsx is commented out, Logout button does not render
 }])
 
 
})
```

**Comment by ImBIOS** [comment]
I'm also facing module build error component testing cookies in cypress
